### ISSUE TITLE
STRINGValueConverter fails to recognize an invalid escape sequence

### ISSUE THREAD

**Issue by waqasilyas** [eclipse/xtext-core#525]
The STRINGValueConverter allows an incomplete escape sequence to be successfully converted to a string. The test below should pass, but it fails:
```
@RunWith(XtextRunner)
@InjectWith(SystemDefinitionTreeInjectorProvider)
class StringTest {
	@Inject
	STRINGValueConverter converter
	
	@Test
	def void test() {
		try {
			val v = converter.toValue('''"\"''', null);
			fail    // Should have had exception
		}
		catch (Exception e) {
			// Should get exception 
		}
	}
}
```
Converting a string `"\\"` should cause an exception, but it returns `"\""` instead. Here is the fix;
```
diff --git a/org.eclipse.xtext/src/org/eclipse/xtext/conversion/impl/STRINGValueConverter.java b/org.eclipse.xtext/src/org/eclipse/xtext/conversion/impl/STRINGValueConverter.java
index d7fe37f..d08f510 100644
--- a/org.eclipse.xtext/src/org/eclipse/xtext/conversion/impl/STRINGValueConverter.java
+++ b/org.eclipse.xtext/src/org/eclipse/xtext/conversion/impl/STRINGValueConverter.java
@@ -61,7 +61,7 @@ public class STRINGValueConverter extends AbstractLexerBasedConverter<String> {
                while (off < end - 1) {
                        aChar = in[off++];
                        if (aChar == '\\') {
-                               if (off < end) {
+                               if (off < end - 1) {
                                        aChar = in[off++];
                                        switch (aChar) {
                                                case 'u':
```

**Comment by cdietrich** [comment]
wont the invalid escape sequence be rejected by the Lexer already?
thus the question is: which grammar do you use?

**Comment by cdietrich** [comment]
and the fix breaks other tests like (from `STRINGConverterTest`)

```
	@Test public void testBrokenStringLiteral_04() throws Exception {
		String s = "' \\z";
		try {
			valueConverter.toValue(s, null);
			fail();
		} catch(ValueConverterWithValueException e) {
			assertEquals(" z", e.getValue());
			assertTrue(e.hasRange());
			assertEquals(2, e.getOffset());
			assertEquals(2, e.getLength());
		}
	}
```

**Comment by waqasilyas** [comment]
Hi,

Thanks for looking into this. Actually my grammar is very forgiving, and allows invalid strings and arbitrary escape sequences so my tests all passed with the change I suggested. Let me correct this and I will try to run the tests on xtext-core myself before sending in a new patch.

It does look like that the test might need to be updated. With new changes the error returned for this would be "string not closed". But the test above is expecting "invalid escape sequence".

I'll get back with a new fix as soon as I can.

**Comment by waqasilyas** [comment]
I am attaching a new patch... this one modifies the above test slightly that was failing. And I also added my test from the issue description.
```
diff --git a/org.eclipse.xtext.tests/src/org/eclipse/xtext/valueconverter/STRINGConverterTest.java b/org.eclipse.xtext.tests/src/org/eclipse/xtext/valueconverter/STRINGConverterTest.java
index 038b8e6..06e325d 100644
--- a/org.eclipse.xtext.tests/src/org/eclipse/xtext/valueconverter/STRINGConverterTest.java
+++ b/org.eclipse.xtext.tests/src/org/eclipse/xtext/valueconverter/STRINGConverterTest.java
@@ -72,9 +72,7 @@ public class STRINGConverterTest extends AbstractXtextTests {
 			fail();
 		} catch(ValueConverterWithValueException e) {
 			assertEquals(" z", e.getValue());
-			assertTrue(e.hasRange());
-			assertEquals(2, e.getOffset());
-			assertEquals(2, e.getLength());
+			assertFalse(e.hasRange());
 		}
 	}
 	
@@ -88,6 +86,17 @@ public class STRINGConverterTest extends AbstractXtextTests {
 			assertFalse(e.hasRange());
 		}
 	}
+	
+	@Test public void testBrokenStringLiteral_06() throws Exception {
+		String s = "\\";
+		try {
+			valueConverter.toValue(s, null);
+			fail();
+		} catch(ValueConverterWithValueException e) {
+			assertEquals("", e.getValue());
+			assertFalse(e.hasRange());
+		}
+	}
 
 	@Test public void testEscapeChars() throws Exception {
 		String s = "\"\\t\\n\\r\\f\\b\"";
diff --git a/org.eclipse.xtext/src/org/eclipse/xtext/conversion/impl/STRINGValueConverter.java b/org.eclipse.xtext/src/org/eclipse/xtext/conversion/impl/STRINGValueConverter.java
index d7fe37f..a338853 100644
--- a/org.eclipse.xtext/src/org/eclipse/xtext/conversion/impl/STRINGValueConverter.java
+++ b/org.eclipse.xtext/src/org/eclipse/xtext/conversion/impl/STRINGValueConverter.java
@@ -61,7 +61,7 @@ public class STRINGValueConverter extends AbstractLexerBasedConverter<String> {
 		while (off < end - 1) {
 			aChar = in[off++];
 			if (aChar == '\\') {
-				if (off < end) {
+				if (off < end - 1) {
 					aChar = in[off++];
 					switch (aChar) {
 						case 'u':
@@ -153,7 +153,6 @@ public class STRINGValueConverter extends AbstractLexerBasedConverter<String> {
 						errorIndex = off - 1;
 						errorLength = 1;
 					}
-					out[outLen++] = aChar;
 				}
 			} else {
 				out[outLen++] = aChar;
```

**Comment by waqasilyas** [comment]
Another patch, slightly different. Giving preference to escape sequence error, over not closed error
```
diff --git a/org.eclipse.xtext.tests/src/org/eclipse/xtext/valueconverter/STRINGConverterTest.java b/org.eclipse.xtext.tests/src/org/eclipse/xtext/valueconverter/STRINGConverterTest.java
index 038b8e6..d19f421 100644
--- a/org.eclipse.xtext.tests/src/org/eclipse/xtext/valueconverter/STRINGConverterTest.java
+++ b/org.eclipse.xtext.tests/src/org/eclipse/xtext/valueconverter/STRINGConverterTest.java
@@ -74,7 +74,7 @@ public class STRINGConverterTest extends AbstractXtextTests {
 			assertEquals(" z", e.getValue());
 			assertTrue(e.hasRange());
 			assertEquals(2, e.getOffset());
-			assertEquals(2, e.getLength());
+			assertEquals(1, e.getLength());
 		}
 	}
 	
@@ -85,6 +85,19 @@ public class STRINGConverterTest extends AbstractXtextTests {
 			fail();
 		} catch(ValueConverterWithValueException e) {
 			assertEquals(" z ", e.getValue());
+			assertTrue(e.hasRange());
+			assertEquals(2, e.getOffset());
+			assertEquals(2, e.getLength());
+		}
+	}
+	
+	@Test public void testBrokenStringLiteral_06() throws Exception {
+		String s = "\\";
+		try {
+			valueConverter.toValue(s, null);
+			fail();
+		} catch(ValueConverterWithValueException e) {
+			assertEquals("", e.getValue());
 			assertFalse(e.hasRange());
 		}
 	}
diff --git a/org.eclipse.xtext/src/org/eclipse/xtext/conversion/impl/STRINGValueConverter.java b/org.eclipse.xtext/src/org/eclipse/xtext/conversion/impl/STRINGValueConverter.java
index d7fe37f..5124446 100644
--- a/org.eclipse.xtext/src/org/eclipse/xtext/conversion/impl/STRINGValueConverter.java
+++ b/org.eclipse.xtext/src/org/eclipse/xtext/conversion/impl/STRINGValueConverter.java
@@ -61,7 +61,7 @@ public class STRINGValueConverter extends AbstractLexerBasedConverter<String> {
 		while (off < end - 1) {
 			aChar = in[off++];
 			if (aChar == '\\') {
-				if (off < end) {
+				if (off < end - 1) {
 					aChar = in[off++];
 					switch (aChar) {
 						case 'u':
@@ -153,7 +153,6 @@ public class STRINGValueConverter extends AbstractLexerBasedConverter<String> {
 						errorIndex = off - 1;
 						errorLength = 1;
 					}
-					out[outLen++] = aChar;
 				}
 			} else {
 				out[outLen++] = aChar;
@@ -174,10 +173,6 @@ public class STRINGValueConverter extends AbstractLexerBasedConverter<String> {
 					} else {
 						errorMessage = getStringNotClosedMessage();
 					}
-				} else {
-					errorMessage = getStringNotClosedMessage();
-					errorIndex = -1;
-					errorLength = -1;
 				}
 			}
 		}

```

**Comment by cdietrich** [comment]
could you provide a pull request?

**Comment by szarnekow** [comment]
Please keep in mind that existing test cases should usually stay untouched to avoid regressions. But as Christian pointed out, a PR would make it tremendously easier to assess the impact of your requested changes.

**Comment by waqasilyas** [comment]
I have created a pull request:
https://github.com/eclipse/xtext-core/pull/526

**Comment by cdietrich** [comment]
Fixed in 2.15
